### ISSUE TITLE
[BlenderBIM] In Gant Chart html to format new option Hour?

### ISSUE THREAD

**Issue by Martin15135215** [IfcOpenShell/IfcOpenShell#2772]
So when we create an gant chart with BlenderBIM

![image](https://user-images.githubusercontent.com/110968398/218792648-cdc98c6f-658c-434a-85d1-14752edaeb31.png)

![image](https://user-images.githubusercontent.com/110968398/218792942-3af3b96c-5cb1-401d-9531-a05e6e1fd6b2.png)

There are different option in the html file for Day, Week, Month and Quarter -> is it possible to add the Hour format?

**Comment by Moult** [comment]
A note that our date calcs currently don't go down to the detail of individual hours, I believe we stop at days. So there is quite a bit of annoying date parsing calcs to support this function.

**Comment by Moult** [comment]
Out of curiosity, when would you ever need the hour option? I've never come across hour by hour planning in the gantt chart in the industry.

**Comment by Plae-Blue-Dot** [comment]
> the hour option? I've never come across hour by hour planning in the gantt chart in the industry.

We've used an hour gantt chart for 24hour job, when there are multiple shifts and people are changing frequently, it's not common but this is one instance where it's applicable

**Comment by Martin15135215** [comment]
> A note that our date calcs currently don't go down to the detail of individual hours, I believe we stop at days. So there is quite a bit of annoying date parsing calcs to support this function.

yeah quite understandable, dates are tricky

**Comment by Martin15135215** [comment]
> Out of curiosity, when would you ever need the hour option? I've never come across hour by hour planning in the gantt chart in the industry.

I can give one from the railway industry

Below is the swiss InterCity connenctions with Train

![](https://upload.wikimedia.org/wikipedia/commons/7/70/IC-Netz_SBB_2017-2020.svg)

So between Geneve and Lausanne there are each day thousend of people communitng with train. Because of that, that's really difficult to do an construction site at the railway -> the solution now is to do night shifts (like bewettn 11pm and 5am) however each time the shifts end, the railway needs to be drivable for the trains.

However the sleepers for example needs to be changed after 10 years -> you have over 3000 km railway -> so you need to do like 200 km each year, changing the sleepers at night -> so there are like small construction site all over swiss that only have a duration like 1-3 month -> with just night shifts under weekdays...

**Comment by Martin15135215** [comment]
> Out of curiosity, when would you ever need the hour option? I've never come across hour by hour planning in the gantt chart in the industry.

I can give one from the railway industry

Below is the swiss InterCity connenctions with Train

![](https://upload.wikimedia.org/wikipedia/commons/7/70/IC-Netz_SBB_2017-2020.svg)

So between Geneve and Lausanne there are each day thousend of people communitng with train. Because of that, that's really difficult to do an construction site at the railway -> the solution now is to do night shifts (like bewettn 11pm and 5am) however each time the shifts end, the railway needs to be drivable for the trains.

However the sleepers for example needs to be changed after 10 years -> you have over 3000 km railway -> so you need to do like 200 km each year, changing the sleepers at night -> so there are like small construction site all over swiss that only have a duration like 1-3 month -> with just night shifts under weekdays...

**Comment by Moult** [comment]
Makes sense (note you seem to have double posted so I deleted the duplicate). The problems are 1) complexity of hourly checks, 2) "day" assumptions need to become dynamic, e.g. is P1D 8 hours, or is P1D "however many hours there are in a day", and how does this work with midnight shifts e.g. 11pm to 5am is 1 day? 2 days? and 3) efficiency of calendar applicability checks. Right now we have "is working day" vs "is holiday" checks, but now we need to check "hey, am I working at 9am? 10? 11? 12? 13? 14?" etc. It's a lot of checks to make sure the entire duration is valid and we've got to make sure code runs fast.

**Comment by Martin15135215** [comment]
So in IfcTaskTime https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcTaskTime.htm
there are the attribute ScheduleStart and ScheduleFinish -> can't we just use that to explitily state the start and end instead of using P1D

BTW what is P1D -> in IfcTaskTime in the Attrubute i only saw https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcDateTime.htm and https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcDuration.htm <- and in here P stands for years

**Comment by Moult** [comment]
P1D is the ISO way of saying a duration (period) of 1 day. P2Y means a period of 2 years.

Some people will tell you that P1D = PT24H, but they are clearly not construction planners who might say that P1D = PT8H.

You can indeed explicitly set the start and end but schedulers expect to be able to type in a duration, and the start/end auto-adjust based on the duration. This duration also has to cascade.

Also when dealing with imports / exports from MS Project, P6, Asta Powerproject they all have assumptions about hours and days. For example for one of them I recall Monday 5pm == Tuesday 8am. Is it right? Is it wrong? Who knows. Basically I got a huge headache and decided to stop at a day.

**Comment by Martin15135215** [comment]
So i have not any files from MS Project, P6, Asta Powerproject. I have the task in a excel sheet...
with the schedule start and schedule end in an column and wanted the create based on the excel sheet, ifctask with ifctasktime with the help from ifcopenshell-python, i think that is possible

**Comment by Martin15135215** [comment]
So in the code example [here ](https://blenderbim.org/docs-python/ifcopenshell-python/code_examples.html#create-a-work-schedule-constructing-a-building-floor-by-floor) states P1W

![image](https://user-images.githubusercontent.com/110968398/219588175-afcd1821-c5da-4c2a-bf33-378e19161392.png)

However in the sample file from bSI

https://github.com/buildingSMART/Sample-Test-Files/blob/8b9eebd5248b72ba7394d90a3e46343fcf338f35/IFC%204.0/BuildingSMARTSpec/construction-scheduling-task.ifc#L265

There is no W -> should i assume that behind the scene the P1W changes to P0Y0M7DT0H0M0S

**Comment by Moult** [comment]
The IFC schema follows the ISO 8601 duration, as do we: https://en.wikipedia.org/wiki/ISO_8601#Durations so "P1W" is simply a short form.

I'll take a more detailed look about the error later. Or maybe @myoualid will beat me to it :)
